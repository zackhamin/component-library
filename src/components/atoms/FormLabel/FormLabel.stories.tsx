import { Story } from '@storybook/react';
import { FormLabel, FormLabelProps } from './FormLabel';

export default {
	title: 'atoms/FormLabel',
	component: FormLabel,
};

const Template: Story<FormLabelProps> = (args) => <FormLabel {...args} />;
const defaultProps = {
	children: 'Label',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Required = Template.bind({});
Required.args = {
	...defaultProps,
	isRequired: true,
};
