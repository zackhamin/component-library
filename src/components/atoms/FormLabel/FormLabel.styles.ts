import styled from 'styled-components';
import { typography } from 'Styles';

const StyledFormLabel = styled.label`
	display: inline-block;
	margin-bottom: 0.8rem;

	&.variant-body {
		font-size: ${typography.size.s2}rem;
	}

	&.variant-heading {
		font-size: ${typography.size.s4}rem;
		font-family: ${typography.type.heading};
	}
`;

export { StyledFormLabel };
