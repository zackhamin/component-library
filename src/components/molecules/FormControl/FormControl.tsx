import { ConditionalFormProps } from '@Types/form';
import { createContext } from 'react';

interface FormControlContext {
	name?: string;
	id: string;
	isRequired?: boolean;
	isInvalid?: boolean;
	isSuccess?: boolean;
}

interface BaseFormControlProps extends FormControlContext {
	children: React.ReactNode;
	className?: string;
}

export type FormControlProps = BaseFormControlProps & ConditionalFormProps;

export const FormControlContext = createContext<FormControlContext | undefined>(
	undefined
);

const FormControl = ({
	children,
	name,
	id,
	isInvalid,
	isRequired,
	isSuccess,
	className,
}: FormControlProps) => {
	return (
		<FormControlContext.Provider
			value={{ name, id, isRequired, isInvalid, isSuccess }}
		>
			<div className={className}>{children}</div>
		</FormControlContext.Provider>
	);
};

export { FormControl };
