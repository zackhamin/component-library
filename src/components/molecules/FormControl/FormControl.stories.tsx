import { FormError } from 'Atoms/FormError';
import { FormLabel } from 'Atoms/FormLabel';
import { Story } from '@storybook/react';
import { TextInput } from 'Atoms/TextInput';
import { FormControl, FormControlProps } from './FormControl';

export default {
	title: 'molecules/FormControl',
	component: FormControl,
	argTypes: {
		isSuccess: {
			control: {
				type: 'boolean',
			},
		},
		isInvalid: {
			control: {
				type: 'boolean',
			},
		},
	},
};

const Template: Story<FormControlProps> = (args) => (
	<FormControl {...args}>
		<FormLabel>Label</FormLabel>
		<TextInput />
		<FormError>Error Message</FormError>
	</FormControl>
);
const defaultProps = {
	name: 'testName',
	id: 'testId',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Invalid = Template.bind({});
Invalid.args = {
	...defaultProps,
	isInvalid: true,
};

export const Required = Template.bind({});
Required.args = {
	...defaultProps,
	isRequired: true,
};

export const Success = Template.bind({});
Success.args = {
	...defaultProps,
	isSuccess: true,
};
