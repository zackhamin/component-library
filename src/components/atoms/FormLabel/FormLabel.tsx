import cn from 'classnames';
import { FormControlContext } from 'Molecules/FormControl';
import { StyledFormLabel } from './FormLabel.styles';
import { useContext } from 'react';

export interface FormLabelProps {
	children: React.ReactNode;
	htmlFor?: string;
	isRequired?: boolean;
	variant?: 'body' | 'heading';
}

const FormLabel = ({
	children = 'Label',
	htmlFor,
	isRequired,
	variant = 'heading',
}: FormLabelProps) => {
	const ctx = useContext(FormControlContext);

	return (
		<StyledFormLabel
			className={cn({ [`variant-${variant}`]: variant })}
			htmlFor={ctx?.id || htmlFor}
		>
			{children}
			{(ctx?.isRequired || isRequired) && <span aria-label="required">*</span>}
		</StyledFormLabel>
	);
};

export { FormLabel };
