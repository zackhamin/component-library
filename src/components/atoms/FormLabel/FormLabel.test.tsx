import { FormControl } from 'Molecules/FormControl';
import { FormLabel } from '../FormLabel';
import { render, screen } from '@testing-library/react';

describe('FormLabel', () => {
	it('renders correctly', () => {
		const { container } = render(<FormLabel htmlFor="test">Label</FormLabel>);
		expect(container).toMatchSnapshot();
	});

	it('adds ID, class, type & name to element', () => {
		render(<FormLabel htmlFor="test">Label</FormLabel>);

		const label = screen.getByText('Label');

		expect(label).toHaveAttribute('for', 'test');
	});

	it('should have required asterisk when isRequired', () => {
		render(
			<FormLabel htmlFor="test" isRequired>
				Label
			</FormLabel>
		);

		const required = screen.getByLabelText('required');

		expect(required).toBeInTheDocument();
	});
});

describe('FormLabel within FormControl', () => {
	it('should have required asterisk when isRequired is passed from parent FormControl', () => {
		render(
			<FormControl id="testId" isRequired>
				<FormLabel>Label</FormLabel>
			</FormControl>
		);

		const required = screen.getByLabelText('required');

		expect(required).toBeInTheDocument();
	});

	it('should have correct for attr when ID is passed from parent FormControl', () => {
		render(
			<FormControl id="testId">
				<FormLabel>Label</FormLabel>
				<input type="text" id="testId" />
			</FormControl>
		);

		const label = screen.getByText('Label');

		expect(label).toHaveAttribute('for', 'testId');
	});
});
