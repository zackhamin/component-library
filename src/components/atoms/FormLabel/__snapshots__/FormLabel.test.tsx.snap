// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLabel renders correctly 1`] = `
.c0 {
  display: inline-block;
  margin-bottom: 0.8rem;
}

.c0.variant-body {
  font-size: 1.4rem;
}

.c0.variant-heading {
  font-size: 1.8rem;
  font-family: BrandonGrotesque-Bold,Arial,Helvetica,sans-serif;
}

<div>
  <label
    class="c0 variant-heading"
    for="test"
  >
    Label
  </label>
</div>
`;
